File: broker-plugins/management-http/src/main/java/resources/js/qpid/management/TrustStore.js
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
define(["dojo/_base/lang",
        "dojo/parser",
        "dojo/query",
        "dojo/_base/connect",
        "dijit/registry",
        "qpid/management/store/CertificateGridWidget",
        "dojox/html/entities",
        "qpid/common/updater",
        "qpid/common/util",
        "qpid/management/addStore",
        "dojo/text!showTrustStore.html",
        "dojo/domReady!"],
    function (lang, parser, query, connect, registry, CertificateGridWidget, entities, updater, util, addStore, template)
    {
        function TrustStore(kwArgs)
        {
            this.controller = kwArgs.controller;
            this.modelObj = kwArgs.tabData.modelObject;
            this.management = this.controller.management;
            this.name = this.modelObj.name;
        }

        TrustStore.prototype.getTitle = function ()
        {
            return "TrustStore: " + this.name;
        };

        TrustStore.prototype.open = function (contentPane)
        {
            var that = this;
            this.contentPane = contentPane;

            contentPane.containerNode.innerHTML = template;
            parser.parse(contentPane.containerNode)
                .then(function (instances)
                {

                    that.keyStoreUpdater = new KeyStoreUpdater(that);
                    that.keyStoreUpdater.update(function ()
                    {
                        updater.add(that.keyStoreUpdater);
                    });

                    var deleteTrustStoreButton = query(".deleteStoreButton", contentPane.containerNode)[0];
                    var deleteButtonNode = registry.byNode(deleteTrustStoreButton);
                    connect.connect(deleteButtonNode, "onClick", function (evt)
                    {
                        that.deleteKeyStore();
                    });

                    var editTrustStoreButton = query(".editStoreButton", contentPane.containerNode)[0];
                    var editButtonNode = registry.byNode(editTrustStoreButton);
                    connect.connect(editButtonNode, "onClick", function (evt)
                    {
                        that.management.load(that.modelObj,
                            {
                                actuals: true,
                                excludeInheritedContext: true
                            })
                            .then(function (data)
                            {
                                addStore.setupTypeStore(that.management, "TrustStore", that.modelObj);
                                addStore.show(data, that.keyStoreUpdater.trustStoreData);
                            }, util.xhrErrorHandler);
                    });

                    var gridNode = query(".managedCertificatesGrid", contentPane.containerNode)[0];
                    that.certificatesGrid = new CertificateGridWidget({
                        management: that.management,
                        modelObj: that.modelObj
                    }, gridNode);
                    that.certificatesGrid.startup();
                    that.certificatesGrid.on("certificatesUpdated", lang.hitch(this, function()
                    {
                        that.keyStoreUpdater.update();
                    }));
                });
        };

        TrustStore.prototype.close = function ()
        {
            updater.remove(this.keyStoreUpdater);
            this.certificatesGrid.destroy();
        };

        function KeyStoreUpdater(tabObject)
        {
            var that = this;
            var containerNode = tabObject.contentPane.containerNode;
            this.keyStoreDetailsContainer = query(".typeFieldsContainer", containerNode)[0];
            this.management = tabObject.management;
            this.modelObj = tabObject.modelObj;
            this.tabObject = tabObject;
            this.contentPane = tabObject.contentPane;

            function findNode(name)
            {
                return query("." + name, containerNode)[0];
            }

            function storeNodes(names)
            {
                for (var i = 0; i < names.length; i++)
                {
                    that[names[i]] = findNode(names[i]);
                }
            }

            storeNodes(["name", "type", "state", "exposedAsMessageSource", "trustAnchorValidityEnforced",
                "certificateRevocationCheckEnabled", "certificateRevocationCheckOfOnlyEndEntityCertificates",
                "certificateRevocationCheckWithPreferringCertificateRevocationList",
                "certificateRevocationCheckWithNoFallback", "certificateRevocationCheckWithIgnoringSoftFailures",
                "certificateRevocationListUrl"]);

        }

        KeyStoreUpdater.prototype.updateHeader = function ()
        {
            this.name.innerHTML = entities.encode(String(this.trustStoreData["name"]));
            this.type.innerHTML = entities.encode(String(this.trustStoreData["type"]));
            this.state.innerHTML = entities.encode(String(this.trustStoreData["state"]));
            this.exposedAsMessageSource.innerHTML =
                entities.encode(String(this.trustStoreData["exposedAsMessageSource"]));
            this.trustAnchorValidityEnforced.innerHTML =
                entities.encode(String(this.trustStoreData["trustAnchorValidityEnforced"]));
            this.certificateRevocationCheckEnabled.innerHTML =
                entities.encode(String(this.trustStoreData["certificateRevocationCheckEnabled"]));
            this.certificateRevocationCheckOfOnlyEndEntityCertificates.innerHTML =
                entities.encode(String(this.trustStoreData["certificateRevocationCheckOfOnlyEndEntityCertificates"]));
            this.certificateRevocationCheckWithPreferringCertificateRevocationList.innerHTML =
                entities.encode(String(this.trustStoreData["certificateRevocationCheckWithPreferringCertificateRevocationList"]));
            this.certificateRevocationCheckWithNoFallback.innerHTML =
                entities.encode(String(this.trustStoreData["certificateRevocationCheckWithNoFallback"]));
            this.certificateRevocationCheckWithIgnoringSoftFailures.innerHTML =
                entities.encode(String(this.trustStoreData["certificateRevocationCheckWithIgnoringSoftFailures"]));
            this.certificateRevocationListUrl.innerHTML = this.trustStoreData["certificateRevocationListUrl"] ?
                entities.encode(String(this.trustStoreData["certificateRevocationListUrl"])) : "";
        };

        KeyStoreUpdater.prototype.update = function (callback)
        {
            if (!this.contentPane.selected && !callback)
            {
                return;
            }

            var that = this;
            this.management.load(this.modelObj, {excludeInheritedContext: true})
                .then(function (data)
                {
                    that.trustStoreData = data;
                    that.updateHeader();
                    that.tabObject.certificatesGrid.update(that.trustStoreData.certificateDetails);

                    if (callback)
                    {
                        callback();
                    }

                    if (that.details)
                    {
                        that.details.update(that.trustStoreData);
                    }
                    else
                    {
                        var implementsManagedInterface = that.management.metadata.implementsManagedInterface("TrustStore",
                            that.trustStoreData.type,
                            "MutableCertificateTrustStore");
                        that.tabObject.certificatesGrid.enableCertificateControls(implementsManagedInterface);

                        require(["qpid/management/store/" + encodeURIComponent(that.trustStoreData.type.toLowerCase())
                                 + "/show"], function (DetailsUI)
                        {
                            that.details = new DetailsUI({
                                containerNode: that.keyStoreDetailsContainer,
                                parent: that
                            });
                            that.details.update(that.trustStoreData);
                        });
                    }
                }, function (error)
                {
                    util.tabErrorHandler(error, {
                        updater: that,
                        contentPane: that.tabObject.contentPane,
                        tabContainer: that.tabObject.controller.tabContainer,
                        name: that.modelObj.name,
                        category: "Trust Store"
                    });
                });
        };

        TrustStore.prototype.deleteKeyStore = function ()
        {
            if (confirm("Are you sure you want to delete trust store '" + this.name + "'?"))
            {
                var that = this;
                this.management.remove(this.modelObj)
                    .then(function (data)
                    {
                        that.contentPane.onClose();
                        that.controller.tabContainer.removeChild(that.contentPane);
                        that.contentPane.destroyRecursive();
                        that.close();
                    }, util.xhrErrorHandler);
            }
        };

        return TrustStore;
    });

File: broker-core/src/main/java/org/apache/qpid/server/model/TrustStore.java
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
package org.apache.qpid.server.model;

import java.security.GeneralSecurityException;
import java.security.cert.Certificate;
import java.util.List;

import javax.net.ssl.TrustManager;

import org.apache.qpid.server.security.CertificateDetails;

@ManagedObject( defaultType = "FileTrustStore" )
public interface TrustStore<X extends TrustStore<X>> extends ConfiguredObject<X>
{
    String TRUST_ANCHOR_VALIDITY_ENFORCED = "trustAnchorValidityEnforced";

    String CERTIFICATE_EXPIRY_WARN_PERIOD = "qpid.truststore.certificateExpiryWarnPeriod";

    @ManagedContextDefault(name = CERTIFICATE_EXPIRY_WARN_PERIOD,
            description = "The number of days before a certificate's expiry"
                          + " that certificate expiration warnings will be written to the log")
    int DEFAULT_CERTIFICATE_EXPIRY_WARN_PERIOD = 30;

    String CERTIFICATE_EXPIRY_CHECK_FREQUENCY = "qpid.truststore.certificateExpiryCheckFrequency";

    @ManagedContextDefault(name = CERTIFICATE_EXPIRY_CHECK_FREQUENCY,
            description = "Period (in days) with which the Broker will repeat the certificate expiration warning")
    int DEFAULT_CERTIFICATE_EXPIRY_CHECK_FREQUENCY = 1;
    @ManagedContextDefault(name = "qpid.truststore.trustAnchorValidityEnforced")
    boolean DEFAULT_TRUST_ANCHOR_VALIDITY_ENFORCED = false;

    String CERTIFICATE_REVOCATION_CHECK_ENABLED = "certificateRevocationCheckEnabled";
    String CERTIFICATE_REVOCATION_CHECK_WITH_IGNORING_SOFT_FAILURES =
            "certificateRevocationCheckWithIgnoringSoftFailures";
    String CERTIFICATE_REVOCATION_CHECK_WITH_PREFERRING_CERTIFICATE_REVOCATION_LIST =
            "certificateRevocationCheckWithPreferringCertificateRevocationList";
    String CERTIFICATE_REVOCATION_CHECK_WITH_NO_FALLBACK = "certificateRevocationCheckWithNoFallback";
    String CERTIFICATE_REVOCATION_CHECK_OF_ONLY_END_ENTITY_CERTIFICATES =
            "certificateRevocationCheckOfOnlyEndEntityCertificates";
    String CERTIFICATE_REVOCATION_LIST_URL = "certificateRevocationListUrl";

    @Override
    @ManagedAttribute(immutable = true)
    String getName();

    @ManagedAttribute(defaultValue = "false", description = "If true the Trust Store will expose its certificates as a special artificial message source.")
    boolean isExposedAsMessageSource();

    @ManagedAttribute(defaultValue = "[]", description = "If 'exposedAsMessageSource' is true, the trust store will expose its certificates only to VirtualHostNodes in this list or if this list is empty to all VirtualHostNodes who are not in the 'excludedVirtualHostNodeMessageSources' list." )
    List<VirtualHostNode<?>> getIncludedVirtualHostNodeMessageSources();

    @ManagedAttribute(defaultValue = "[]", description = "If 'exposedAsMessageSource' is true and 'includedVirtualHostNodeMessageSources' is empty, the trust store will expose its certificates only to VirtualHostNodes who are not in this list." )
    List<VirtualHostNode<?>> getExcludedVirtualHostNodeMessageSources();

    @ManagedAttribute(defaultValue = "${qpid.truststore.trustAnchorValidityEnforced}",
                       description = "If true, the trust anchor's validity dates will be enforced.")
    boolean isTrustAnchorValidityEnforced();

    @ManagedAttribute(defaultValue = "false", description = "If true, enable certificates revocation.")
    boolean isCertificateRevocationCheckEnabled();

    @ManagedAttribute(defaultValue = "false", description = "If true, check the revocation status of only end-entity certificates.")
    boolean isCertificateRevocationCheckOfOnlyEndEntityCertificates();

    @ManagedAttribute(defaultValue = "true", description = "If true, prefer CRL (specified in certificate distribution points) to OCSP, if false prefer OCSP to CRL.")
    boolean isCertificateRevocationCheckWithPreferringCertificateRevocationList();

    @ManagedAttribute(defaultValue = "true", description = "If true, disable fallback to CRL/OCSP (if 'certificateRevocationCheckWithPreferringCertificateRevocationList' set to true, disable fallback to OCSP, otherwise disable fallback to CRL in certificate distribution points).")
    boolean isCertificateRevocationCheckWithNoFallback();

    @ManagedAttribute(defaultValue = "false", description = "If true, revocation check will succeed if CRL/OCSP response cannot be obtained because of network error or OCSP responder returns internalError or tryLater.")
    boolean isCertificateRevocationCheckWithIgnoringSoftFailures();

    @ManagedAttribute(oversize = true, description = "If set, certificates will be validated only against CRL file (CRL in distribution points and OCSP will be ignored).", oversizedAltText = OVER_SIZED_ATTRIBUTE_ALTERNATIVE_TEXT)
    String getCertificateRevocationListUrl();

    @DerivedAttribute
    String getCertificateRevocationListPath();

    @DerivedAttribute(description = "List of details about the certificates like validity dates, SANs, issuer and subject names, etc.")
    List<CertificateDetails> getCertificateDetails();

    @DerivedAttribute
    int getCertificateExpiryWarnPeriod();

    @DerivedAttribute
    int getCertificateExpiryCheckFrequency();

    TrustManager[] getTrustManagers() throws GeneralSecurityException;

    Certificate[] getCertificates() throws GeneralSecurityException;

}
